bot/services/class_service.py

```diff
@@ -0,0 +1,433 @@
+        """
+        Gets an archive category that has room for at least one channel.
+        If None is returned, there are no archive categories that are NOT full.
+        """
+        archive_categories = bot_secrets.secrets.class_archive_category_ids
+        for category in self.bot.guild.categories:
+            if category.id not in archive_categories:
+                continue
+            if len(category.channels) < MAX_CHANNELS_PER_CATEGORY:
+                return category
+        return None
+
+    async def _get_or_create_category(self, cls: ClassChannelScaffold) -> CategoryChannel:
+        """
+        Gets or creates a new category channel for the given scaffold.
+        """
+        category_name = cls.intended_category
+        for category in self.bot.guild.categories:
+            if category_name == category.name:
+                return category
+        return await self.bot.guild.create_category(name=category_name)
+
+    async def _get_or_create_role(self, cls: ClassChannelScaffold) -> discord.Role:
+        """
+        Gets or creates a new role for the given class scaffold.
+        """
+        for r in self.bot.guild.roles:
+            if r.name == cls.class_code:
+                return r
+        return await self.bot.guild.create_role(
+            name=cls.class_code,
+            mentionable=True,
+            reason=f'Class creation or could not find class role {cls.class_code}'
+        )
+
+    async def _sync_perms(self, cls: ClassChannel) -> None:
+        """
+        Syncs the permissions for the given class channel.
+        This method is specifically for class channels that are being created, inserted, or unarchived.
+
+        The following permissions are set, for the given roles:
+        CLASS ROLE      POSITION 3      VIEW_CHANNEL = TRUE
+        CLEANUP ROLE    POSITION 2      VIEW_CHANNEL = FALSE
+        """
+        role = await self._get_or_create_role(cls)
+        cleanup = await self._get_or_create_cleanup()
+        if not (channel := self.bot.guild.get_channel(cls.channel_id)):
+            await self._send_failure(cls, 'Syncing Perms Failed', f'The channel `{cls.channel_id}` does not exist.')
+            return
+        await channel.set_permissions(role, view_channel=True)
+        await channel.set_permissions(cleanup, view_channel=False)
+        await role.edit(mentionable=True, position=3)
+        await cleanup.edit(position=2)
+
+    async def _get_or_create_cleanup(self) -> discord.Role:
+        """
+        Fetches the 'Cleanup' role from the guild.
+        If not found, a new role with the name 'Cleanup' is created and returned.
+        """
+        for role in self.bot.guild.roles:
+            if role.name == 'Cleanup':
+                return role
+        return await self.bot.guild.create_role(name='Cleanup')
+
+    async def _send_welcome(self, cls: ClassChannel, just_created: bool, user: discord.User):
```

Comment:
The welcome should also ping the user that created it so that they can find the channel